src/Widgets/MediaMenu.vala

```diff
@@ -0,0 +1,327 @@
+                append (contractor_entry);
+                append (rate_media);
+                append (new Gtk.SeparatorMenuItem ());
+                append (queue_media);
+                append (new Gtk.SeparatorMenuItem ());
+                append (remove_media);
+                append (import_to_library);
+                remove_media.label = _("Remove from Device");
+                break;
+            case ViewWrapper.Hint.READ_ONLY_PLAYLIST:
+                append (scroll_to_current);
+                append (new Gtk.SeparatorMenuItem ());
+                append (file_browse);
+                append (contractor_entry);
+                if (generic_list.playlist == App.player.queue_playlist) {
+                    append (new Gtk.SeparatorMenuItem ());
+                    append (remove_media);
+                    remove_media.label = _("Remove from Queue");
+                }
+                break;
+            case ViewWrapper.Hint.SMART_PLAYLIST:
+                append (scroll_to_current);
+                append (new Gtk.SeparatorMenuItem ());
+                append (edit_media);
+                append (file_browse);
+                append (contractor_entry);
+                append (rate_media);
+                append (new Gtk.SeparatorMenuItem ());
+                append (queue_media);
+                append (add_to_playlist);
+                break;
+        }
+
+        show_all ();
+
+        scroll_to_current.activate.connect (() => {
+            generic_list.scroll_to_current_media (true);
+        });
+
+        App.player.playback_stopped.connect (() => {
+            scroll_to_current.sensitive = false;
+        });
+
+        App.player.playback_started.connect (() => {
+            scroll_to_current.sensitive = true;
+        });
+
+        edit_media.activate.connect (edit_media_clicked);
+        file_browse.activate.connect (file_browse_clicked);
+        import_to_library.activate.connect (import_to_library_clicked);
+        queue_media.activate.connect (queue_clicked);
+        rate_media.activate.connect (rate_media_clicked);
+        remove_media.activate.connect (remove_media_clicked);
+    }
+
+    public void popup_media_menu (Gee.Collection<Media> selection) {
+        var media_menu_new_playlist = new Gtk.MenuItem.with_label (_("New Playlist…"));
+        media_menu_new_playlist.activate.connect (media_menu_new_playlist_clicked);
+
+        var add_to_playlist_menu = new Gtk.Menu ();
+        add_to_playlist_menu.append (media_menu_new_playlist);
+
+        if (generic_list.parent_wrapper.library.support_playlists () == false) {
+            media_menu_new_playlist.visible = false;
+        }
+        foreach (var playlist in generic_list.parent_wrapper.library.get_playlists ()) {
+            // Don't include this playlist in the list of available options
+            if (playlist == generic_list.playlist)
+                continue;
+
+            if (playlist.read_only == true)
```

Comment:
Maybe add braces here while we are at it? And this condition and the above may be combined.